### ISSUE TITLE
iPhone X Issues Megathread

### ISSUE THREAD

**Issue by chrisballinger** [jessesquires/JSQMessagesViewController#2179]
Although this library is deprecated, there are a large number of people still using it in production who will probably all be rushing to solve iPhone X layout issues over the next few weeks. Since the [network graph](https://github.com/jessesquires/JSQMessagesViewController/network) is too large to display on GitHub, it's hard to see what other people have done to address these issues. As the legacy maintainer I want to make sure we can still maintain a dialogue for certain critical bugs.

The main issue that I've found is that the constraints for the input toolbar needed to be adjusted for the "safe area" (see example below). This can be solved by adding "safe area" support to the xib file, but it requires a deployment target of iOS 9 or higher. ~~I also noticed that although it solved the issue in the demo project, it was still broken in my own project (and still haven't figured out why), so your milage may vary.~~ (just kidding the old xib was cached)

I'll soon be creating a new `iphone_x` branch containing my fixes so far. Feel free to discuss your solutions here, but please keep conversation constructive and limited to iPhone X issues only.

Additional Reading:
* https://developer.apple.com/ios/human-interface-guidelines/overview/iphone-x/
* https://developer.apple.com/documentation/uikit/uiview/positioning_content_relative_to_the_safe_area
* https://github.com/MessageKit/MessageKit/blob/9ff1a122c33231d88a4654e1b11b85e1bb0ccf08/Sources/Controllers/MessagesViewController.swift#L130

Before:
<img width="545" alt="screen shot 2017-11-01 at 2 22 54 pm" src="https://user-images.githubusercontent.com/518687/32299102-8c33be22-bf11-11e7-8360-a7fda4b267b5.png">
After:
<img width="545" alt="screen shot 2017-11-01 at 2 25 47 pm" src="https://user-images.githubusercontent.com/518687/32299463-eb391060-bf12-11e7-9021-5895f2043932.png">

**Comment by chrisballinger** [comment]
Feel free to cherry-pick this commit: https://github.com/ChatSecure/JSQMessagesViewController/commit/bec4c848bb59c65ae040e1c31f2e8052b97657e2

**Comment by otymartin** [comment]
Beautiful, thanks for this @chrisballinger and really good idea opening a consolidated iPhone X issue page.

**Comment by chrisballinger** [comment]
@otymartin I was able to get it working with a clean build using just the xib modifications in my patch. Your patch probably works better for those that still need iOS 7/8 support.

**Comment by otymartin** [comment]
@chrisballinger curious if you get the issue where theres clear space between the bottom of the toolbar and the bottom of the screen allowing message bubbles to be seen through?
like so.

<img width="311" alt="screen shot 2017-11-02 at 10 29 45 pm" src="https://user-images.githubusercontent.com/2222881/32363685-df3176a6-c046-11e7-91b2-76f06266ee1c.png">

**Comment by chrisballinger** [comment]
Hmmm no but I realized that the constraints for the xib method only need to be pinned to the bottom safe area. (the translucent bleed through in the navbar is clipped at the moment)

**Comment by otymartin** [comment]
@chrisballinger question are you using the master or develop branch?

**Comment by eliburke** [comment]
@chrisballinger I cherry picked your bec4c84 commit for JSQMessagesViewController.xib and it fixed the bottom margin. 

I also needed to use @lucytheslayer's fix from #2172 to fix the top cell underlapping the Nav Bar. 

````
// Swift
if #available(iOS 11.0, *){
    self.collectionView.contentInsetAdjustmentBehavior = .never 
    self.collectionView.contentInset = UIEdgeInsetsMake(64, 0, 0, 0)
    self.collectionView.scrollIndicatorInsets = self.collectionView.contentInset 
}

// ObjC
if (@available(iOS 11.0, *)) {
    self.collectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    self.collectionView.contentInset = UIEdgeInsetsMake(64, 0, 0, 0);
    self.collectionView.scrollIndicatorInsets = self.collectionView.contentInset;
}
````

The final change I had to make was to set self.inputToolbar.barTintColor to match self.inputToolbar.contentView.backgroundColor. Without it, the home indicator was choosing a gray shade that matches nothing else in my app.

**Comment by chrisballinger** [comment]
@otymartin I maintain [my own fork](https://github.com/ChatSecure/JSQMessagesViewController/tree/7.3.4-chatsecure) that's mostly 7.3.4 with some cherry-picks from develop and a few other small modifications. At one point a while ago I tried to migrate my project to 8.0.0/develop and had a number of issues so I had to revert. I'm not sure what to do with the current state of upstream because IMO both the master branch and develop are broken in their own ways.

@eliburke Thanks for that tip! Is there a way to make that `64` not just a magic number?

**Comment by SD10** [comment]
@eliburke I'm not sure #2172 is a fix because the `contentInset` is hard coded. It also needs to consider nav bar variations + JSQ's `additionalContentInset` property which has a corresponding set method.

**Comment by eliburke** [comment]
Good point. I was rushing and didn’t pay attention to anything besides the fact that it appeared to fix the (obvious) underlap. 

Worse than magic values, it looks to be wrong on iPhone X— there is too much inset when the keyboard is closed, and it still underlaps when open.

**Comment by Tulleb** [comment]
@chrisballinger I'm afraid your commit isn't enough for those who wants to fix it from the original JSQ developer branch.

@jessesquires was doing some magic removing the `inputToolbar` which makes the .xib file constraint quite useless.
You don't seem to have [this line](https://github.com/jessesquires/JSQMessagesViewController/blob/c247626cdb31724ae706733ce7bd8eb4bd2d2513/JSQMessagesViewController/Controllers/JSQMessagesViewController.m#L162) in your own project, and adapting the current Jesse's current develop branch to yours might be too much work for most of us.

I am taking a look on how to fix this from the original develop branch since the `didMoveToWindow` fix isn't satisfying enough for me.

I'm looking for the place where the final `inputToolbar` is added back to the view after this `removeFromSuperview`...

**Comment by Tulleb** [comment]
I am proposing a fixed fork based on the JSQ latest `develop` branch commit.

It is using the `didMoveToWindow` solution. Not ideal but worth to try while waiting for Apple's answer about `inputAccessoryView`'s safe area layout guide attachment, or any other better fix.

You can add this to your Podfile, replacing the previous JSQ line:

    pod 'JSQMessagesViewController', :git => 'https://github.com/Tulleb/JSQMessagesViewController.git', :branch => 'develop', :inhibit_warnings => true

**Comment by chrisballinger** [comment]
@Tulleb I don't mind merging the right fix(es) into upstream/develop once they're ready.

**Comment by Tulleb** [comment]
@chrisballinger I spent my evening on it yesterday. I gave up after a few hours and resigned to use the `didMoveWindow` solution.

The problem is that the `inputToolbar` is not great to use as an `inputAccessoryView`. In my opinion we would need to convert it into a `UIView` in order to fix it correctly. But from what I read, it seems to be a problem from Apple's side. So maybe a fix would arrive in a further iOS version.

@chrisballinger do you want me to open a PR with the `didMoveToWindow` feature while we get something better?

**Comment by alessiettojr** [comment]
Hello everyone, I'm new to the game and wanted to know if there was a guide to fixing the view problem for iphone X.
Thank you all

**Comment by SD10** [comment]
@Tulleb `didMoveToWindow` is what we decided to use at [MessageKit](https://github.com/MessageKit/MessageKit/pull/350/files) as well. I think this is becoming a widely used solution, have had issues trying to solve this problem other ways.

**Comment by Tulleb** [comment]
@SD10 Didn't you try to set up a `UIView` as the `inputAccessoryView` in the first place? I don't understand why it absolutely has to be a `UIToolbar`. I'm not sure this is the appropriate expected behaviour from Apple for this element.

**Comment by Tulleb** [comment]
@alessiettojr You will find your answer in the previous messages of this thread.

**Comment by alessiettojr** [comment]
@otymartin perfect runs smoothly, thank you very much

**Comment by SD10** [comment]
@Tulleb You're right, I forgot about the `UIToolbar` in JSQMVC. Yes, we decided to go with a `UIView` from the start to get the flexibility we wanted

**Comment by J-Bossi** [comment]
For me an extension to the JSQMessagesInputToolbar works

`extension JSQMessagesInputToolbar {
  override open func didMoveToWindow() {
    super.didMoveToWindow()
    if #available(iOS 11.0, *) {
      if self.window?.safeAreaLayoutGuide != nil {
        self.bottomAnchor.constraintLessThanOrEqualToSystemSpacingBelow((self.window?.safeAreaLayoutGuide.bottomAnchor)!, multiplier: 1.0).activate()
      }
    }
  }
}`

https://stackoverflow.com/questions/46886345/jsqmessagesviewcontroller-inputtoolbar-ios-11-iphone-x-swift4/46971214#46971214

**Comment by alanscarpa** [comment]
Using the extension provided by @J-Bossi fixes the input toolbar initial position, but now the most recent message is partially covered by the input toolbar.  Anyone have experience with this?

**Comment by dmathewwws** [comment]
@alanscarpa or @otymartin did you find an answer to your question? im running into the same problem

**Comment by nbonamy** [comment]
Hello,

I am also facing a different issue: once the keyboard has popped out, if you close it (by swiping down) for instance then the very bottom part gets transparent:

![screen shot 2018-03-15 at 18 58 24](https://user-images.githubusercontent.com/956207/37481816-ff9850b4-2882-11e8-8c99-a144ca8ad93a.png)

Any one has a clue how to fix this?

Thanks,
Nicolas

**Comment by Tulleb** [comment]
`pod 'JSQMessagesViewController', :git => 'https://github.com/Tulleb/JSQMessagesViewController.git', :branch => 'develop', :inhibit_warnings => true`

**Comment by otymartin** [comment]
@Tulleb does this fix the transparent keyboard issue @nbonamy mentioned above? curious

**Comment by nbonamy** [comment]
@Tulleb I was already using your fork although not on the 'develop' branch. Unfortunately switching to 'develop' does not fix the issue I am mentionning. The area below the keyboard is OK on view controller display but if you open the keyboard and then close it (by swiping down or posting a message), the area becomes transparent.

**Comment by otymartin** [comment]
@nbonamy @Tulleb I noticed behaviour that causes and removes the transparency. When you type a long enough message that forces the textView to expand to 2 or more lines, the transparancy issue arises. Do the same thing, compose a message with more than 2 lines (textview must expand) but this time press send `while the keyboard is resigned` it resets the layout and transparency is gone. So I dont know objective C meaning i cant trace it but whoever does should be able to trace whatever causes the layout glitch when textView expands to more than 2 lines

**Comment by nbonamy** [comment]
@otymartin @Tulleb in my case it does not even have to expand. just pop-out the keyboard and then slide the collection view down to collapse it. I tried some fixes in `- (void)textViewDidEndEditing:(UITextView *)textView` but did not succeed.

**Comment by Tulleb** [comment]
True, sorry for the disappointment. In my case, I just need to send one message and then pop back: so the glitch isn't a real issue...

I guess it's time to move on to @SD10's https://github.com/MessageKit/MessageKit.

**Comment by SD10** [comment]
I'm pretty sure we solved this issue but we have enough existing problems of our own. JSQMVC is like 4x faster than MessageKit. This doesn't matter all too much if you're managing your data source reasonably, however, I'm still considering a rollback to an API that mirrors JSQMVC 😕

**Comment by otymartin** [comment]
@SD10 you solved in for JSQMVC or MessageKit? Im just holding onto JSQMVC until MessageKit is ready

**Comment by nbonamy** [comment]
`<warning><hack class="ugly">`
Finally "solved" this by adding an opaque UIView that I position at the bottom of the screen in the case the device is an iPhone X. Will do for the moment (once I add device rotation handling...).
`</hack></warning>`

**Comment by SD10** [comment]
Sorry for the lack of clarity @otymartin. Meant MessageKit

**Comment by Luten** [comment]
@nbonamy Found a solution without empty space below input toolbar, the idea that not the whole toolbar should be above the safe area, but only its content:

    -(void) didMoveToWindow{
        [super didMoveToWindow];
        if (@available(iOS 11.0, *)) {
            if (self.window.safeAreaLayoutGuide != nil) {
                [[self.contentView bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:self.window.safeAreaLayoutGuide.bottomAnchor multiplier:1.0].active = YES;
            }
        }
    }

**Comment by Ariandr** [comment]
@Luten 
I created a fork and implemented that solution which fixes empty space. It's based on the release `7.3.5`.

Usage in `Podfile`:

`pod 'JSQMessagesViewController', :git => 'https://github.com/Ariandr/JSQMessagesViewController.git', :branch => 'master', :inhibit_warnings => true`

**Comment by otymartin** [comment]
@Luten @Ariandr 
I tried that solution.
```swift
extension JSQMessagesInputToolbar {
    
    open override func didMoveToWindow() {
        super.didMoveToWindow()
        guard #available(iOS 11.0, *), App.shared.isiPhoneX, let window = self.window else { return }
        self.contentView.bottomAnchor.constraintLessThanOrEqualToSystemSpacingBelow(window.safeAreaLayoutGuide.bottomAnchor, multiplier: 1.0).isActive = true
    }
}
```

But the input textbox layout breaks atleast in simulator. Initially its height is like at zero, and expands as you add text.


![screen shot 2018-05-10 at 12 53 46 pm](https://user-images.githubusercontent.com/2222881/39882303-5e4edfd4-5451-11e8-9d3b-291a93801d2b.png)
![screen shot 2018-05-10 at 12 54 06 pm](https://user-images.githubusercontent.com/2222881/39882305-5e5d8660-5451-11e8-95de-2f03a316004c.png)

**Comment by Ariandr** [comment]
@otymartin 
I think I know your issue. My solution (in fork) is built on the top of the initial solution, not only using that code.

> This can be solved by adding "safe area" support to the xib file, but it requires a deployment target of iOS 9 or higher.

I updated .xib file and target to 9.0. And added that piece of code.
So, feel free to install my custom pod and look at it. :)

`pod 'JSQMessagesViewController', :git => 'https://github.com/Ariandr/JSQMessagesViewController.git', :branch => 'master', :inhibit_warnings => true`

**Comment by otymartin** [comment]
@Ariandr ohh beautiful will take a look now.

Thanks

**Comment by apwelsh** [comment]
@Ariandr your fix does not work.  The keyboard is way too high on the screen, and the text view is not bound by the safe area, resulting in messages appearing below the keyboard.

**Comment by Ariandr** [comment]
@apwelsh Have you installed my fork or just inserted the piece of code?
I don't have such problems in my project using the fork. The only problem is when the keyboard is shown, the text view is not placed precisely above it.

Maybe you have a special setup which causes those problems?

![screenshot at may 29 12-40-55](https://user-images.githubusercontent.com/13591333/40651236-dac573b6-633d-11e8-970b-3a0d094a6678.png)
![screenshot at may 29 12-41-30](https://user-images.githubusercontent.com/13591333/40651240-dd9a453a-633d-11e8-9ee2-cb8d1e8b9ffd.png)
![screenshot at may 29 12-45-59](https://user-images.githubusercontent.com/13591333/40651401-46a4cfa0-633e-11e8-8754-a89c75ef65bc.png)

**Comment by DavidFaraday** [comment]
Any idea how to fix this with iOS12? on iOS 11.4  this worked fine
```
in JSQMessagesInputToolbar.m
    
-(void) didMoveToWindow{
[super didMoveToWindow];
 if (@available(iOS 11.0, *)) {
     [[self bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:self.window.safeAreaLayoutGuide.bottomAnchor multiplier:1.0].active = YES;
     }
}
```

**Comment by apwelsh** [comment]
@chrisballinger is there a merged version of you patch somewhere?  I tried the master at Feel free to cherry-pick this commit: ChatSecure/JSQMessagesViewController@bec4c84 and it is way behind on commits, and does not implement a functional fix for iPhone X.

**Comment by apwelsh** [comment]
So, this is what worked for me,  I went back to the official pod source for JSQMessagesViewController, and then in my code, I create a new Category JSQMessagesInputToolBar+SafeArea

The header file looks like this:
```
//
//  JSQMessagesInputToolbar+SafeArea.h
//  PeerRate
//

#import <JSQMessagesViewController/JSQMessagesViewController.h>

@interface JSQMessagesInputToolbar (SafeArea)

- (void)didMoveToWindow;

@end
```

And the implementation class looks like this
```
//
//  JSQMessagesInputToolbar+SafeArea.m
//

#import "JSQMessagesInputToolbar+SafeArea.h"

@implementation JSQMessagesInputToolbar (SafeArea)

- (void)didMoveToWindow {
    [super didMoveToWindow];
    NSLog(@"didMoveToWindow");
    if (@available(iOS 11.0, *)) {
        
        UILayoutGuide * _Nonnull safeArea = self.window.safeAreaLayoutGuide;
        if (safeArea) {
            NSLayoutYAxisAnchor * _Nonnull bottomAnchor = safeArea.bottomAnchor;
            [[self bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:bottomAnchor multiplier:1.0].active = YES;
        }
    }
}

@end
```

No other changes were required, except to include the JSQMessagesInputToolbar in my ChatViewController.m class that subclasses the JSQMessagesViewController.

**Comment by sebasdeweert** [comment]
This will work: https://stackoverflow.com/a/47810515/2084596

**Comment by LatteCat** [comment]
> @nbonamy Found a solution without empty space below input toolbar, the idea that not the whole toolbar should be above the safe area, but only its content:
> 
> ```
> -(void) didMoveToWindow{
>     [super didMoveToWindow];
>     if (@available(iOS 11.0, *)) {
>         if (self.window.safeAreaLayoutGuide != nil) {
>             [[self.contentView bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:self.window.safeAreaLayoutGuide.bottomAnchor multiplier:1.0].active = YES;
>         }
>     }
> }
> ```

It works! thx

**Comment by Artem85** [comment]
> So, this is what worked for me, I went back to the official pod source for JSQMessagesViewController, and then in my code, I create a new Category JSQMessagesInputToolBar+SafeArea
> 
> The header file looks like this:
> 
> ```
> //
> //  JSQMessagesInputToolbar+SafeArea.h
> //  PeerRate
> //
> 
> #import <JSQMessagesViewController/JSQMessagesViewController.h>
> 
> @interface JSQMessagesInputToolbar (SafeArea)
> 
> - (void)didMoveToWindow;
> 
> @end
> ```
> And the implementation class looks like this
> 
> ```
> //
> //  JSQMessagesInputToolbar+SafeArea.m
> //
> 
> #import "JSQMessagesInputToolbar+SafeArea.h"
> 
> @implementation JSQMessagesInputToolbar (SafeArea)
> 
> - (void)didMoveToWindow {
>     [super didMoveToWindow];
>     NSLog(@"didMoveToWindow");
>     if (@available(iOS 11.0, *)) {
>         
>         UILayoutGuide * _Nonnull safeArea = self.window.safeAreaLayoutGuide;
>         if (safeArea) {
>             NSLayoutYAxisAnchor * _Nonnull bottomAnchor = safeArea.bottomAnchor;
>             [[self bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:bottomAnchor multiplier:1.0].active = YES;
>         }
>     }
> }
> 
> @end
> ```
> No other changes were required, except to include the JSQMessagesInputToolbar in my ChatViewController.m class that subclasses the JSQMessagesViewController.

Did you #import "JSQMessagesInputToolbar+SafeArea.m" in your ChatViewController.m class?

**Comment by PasqualePuzio** [comment]
The solution proposed by @apwelsh worked for me

**Comment by tecworldit** [comment]
> Beautiful, thanks for this @chrisballinger and really good idea opening a consolidated iPhone X issue page.
> @chrisballinger question how did you solve safe-area guide issue in your own project? You mentioned the old xib being cached, I can't seem to solve it in the demo or my own project.
> 
> **Update**
> This persons answer is the only solution that worked for me. Placing this code in `JSQMessagesInputToolbar.m`
> 
> ```swift
> -(void) didMoveToWindow{
> [super didMoveToWindow];
>  if (@available(iOS 11.0, *)) {
>      [[self bottomAnchor] constraintLessThanOrEqualToSystemSpacingBelowAnchor:self.window.safeAreaLayoutGuide.bottomAnchor multiplier:1.0].active = YES;
>      }
> }
> ```
> 
> https://stackoverflow.com/questions/46439975/jsqmessageviewcontroller-ios11-toolbar

Perfect
